### PR TITLE
Add support for attr classes in Dataset

### PR DESCRIPTION
**Pull request by AdrienCorenflos**
See issue: https://github.com/tensorflow/tensorflow/issues/38192
I have not run the tests (can't manage to build tf locally...) but I modified a venv locally and it seems to work on the example attached.

### CODE REVIEW

**Comment by AdrienCorenflos** [comment]
@googlebot I signed it!

File: tensorflow/python/data/util/structure_test.py [hunk 1]
```diff
@@ -748,6 +749,17 @@ def testCustomMapping(self):
...
+  def testAttrClass(self):
+    @attr.s
+    class AttrClass:
+        x = attr.ib()
+        y = attr.ib()
+
+    elem = AttrClass(x=constant_op.constant(1.), y=constant_op.constant(2.))
+    spec = structure.type_spec_from_value(elem)
+    self.assertIsInstance(spec, AttrClass)
+    self.assertEqual(spec.x, tensor_spec.TensorSpec([], dtypes.float32))
```
Comment (aaudiber) [inline tensorflow/python/data/util/structure_test.py:761]: Also check `spec.y`

File: tensorflow/python/data/util/structure_test.py [hunk 2]
```diff
@@ -748,6 +749,17 @@ def testCustomMapping(self):
...
+
```
Comment (aaudiber) [inline tensorflow/python/data/util/structure_test.py:762]: Please add a test for an `ObjectProxy` of `attr.s`, similar to the test for `ObjectProxy` of `namedtuple` below

**Review by aaudiber — CHANGES_REQUESTED** [review]
Thanks for contributing this improvement @AdrienCorenflos. Can you add a test to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/dataset_test.py which tests that `attr.s` classes can be used by `tf.data.Dataset` code?

**Comment by AdrienCorenflos** [comment]
@aaudiber see https://github.com/tensorflow/tensorflow/pull/38266/commits/c4bf9b0c2ef80b48ac76c7128a3b34cda87f8d0a

File: tensorflow/python/data/util/structure_test.py [hunk 3]
```diff
@@ -767,6 +768,19 @@ def testObjectProxy(self):
...
+        x = attr.ib()
+        y = attr.ib()
+
+    elem = AttrClass(x=constant_op.constant(1.), y=constant_op.constant(2.))
+    proxied_elem = wrapt.ObjectProxy(elem)
+    proxied_spec = structure.type_spec_from_value(proxied_elem)
+    self.assertEqual(structure.type_spec_from_value(attr.evolve(elem)),
+                     proxied_spec)
+
+
```
Comment (aaudiber) [inline tensorflow/python/data/util/structure_test.py:783]: nit: extra newline

File: tensorflow/python/data/kernel_tests/dataset_test.py [hunk 4]
```diff
@@ -347,6 +348,18 @@ def testOptionalDatasetSpec(self):
...
+
+    @attr.s
+    class AttrClass:
+      x = attr.ib()
+
+    self._testDatasetSpec(
+        AttrClass(x=constant_op.constant(0)),
+        AttrClass(x=tensor_spec.TensorSpec([], dtypes.int32)))
+
+
```
Comment (aaudiber) [inline tensorflow/python/data/kernel_tests/dataset_test.py:362]: nit: extra newline

**Review by aaudiber — APPROVED** [review]
LGTM, thanks!

**Comment by gbaned** [comment]
@AdrienCorenflos Can you please fix build failures ? Thanks!

**Comment by AdrienCorenflos** [comment]
Hi
How do I know which one of the 1000+ failures and errors is due to my change? There are no logs for errors, just error code 1...

**Comment by jsimsa** [comment]
If you click on Target Log tab you should see this: 
Screen Shot 2020-04-09 at 9 26 07 AM

**Comment by aaudiber** [comment]
@AdrienCorenflos you can find the detailed error message under the "Target Log" tab. It looks like the cause of the failures is "ModuleNotFoundError: No module named 'attr'". This is because the `attr` module may not always be available (TF is still in the process of dropping python2 within Google). You can fix this issue by guarding your attr imports with

```
try:
  import attr  # pylint:disable=g-import-not-at-top
except ImportError:
  attr = None
```

and then checking whether attr is None before using it. See function_test.py for an example of doing this: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/eager/function_test.py

**Comment by gbaned** [comment]
@AdrienCorenflos Can you please check @aaudiber's comments and resolve conflicts?. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by AdrienCorenflos** [comment]
Hi, FYI I have not continued on this one because there is an underlying bug with datasets that needs correcting first (basically you have different behaviours when you do tf.data.Dataset.from_tensor_slices([namedtuples]) vs data.map(namedtuples) which was propagated to my stuff). I will raise a separate issue when I have some time.

**Comment by AdrienCorenflos** [comment]
@gbaned see this
https://github.com/tensorflow/tensorflow/issues/39403

**Pull request closed without merging**